filter_test.go

```diff
@@ -0,0 +1,241 @@
+			expectedNode: equality,
+		},
+		{
+			name:         "and one",
+			filter:       "(&(eq=eqval))",
+			expectedNode: &And{Nodes: []FilterNode{equality}},
+		},
+		{
+			name:         "and many",
+			filter:       "(&(eq=eqval)(present=*))",
+			expectedNode: &And{Nodes: []FilterNode{equality, presence}},
+		},
+		{
+			name:         "or one",
+			filter:       "(|(eq=eqval))",
+			expectedNode: &Or{Nodes: []FilterNode{equality}},
+		},
+		{
+			name:         "or many",
+			filter:       "(|(eq=eqval)(present=*))",
+			expectedNode: &Or{Nodes: []FilterNode{equality, presence}},
+		},
+		{
+			name:         "not",
+			filter:       "(!(present=*))",
+			expectedNode: &Not{Node: presence},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			n, err := Parse(tt.filter)
+			require.NoError(t, err)
+			require.Equal(t, tt.expectedNode, n)
+		})
+	}
+}
+
+func Test_FilterParse_Fail(t *testing.T) {
+	tests := []struct {
+		name   string
+		filter string
+		err    error
+	}{
+		{
+			name:   "empty input",
+			filter: "",
+			err:    ErrUnexpectedEOF,
+		},
+		{
+			name:   "short input",
+			filter: "(",
+			err:    ErrUnexpectedEOF,
+		},
+		{
+			name:   "empty attr name",
+			filter: "(=value)",
+			err:    ErrEmptyAttrName,
+		},
+		{
+			name:   "empty attr value",
+			filter: "(attr=)",
+			err:    ErrEmptyAttrValue,
+		},
+		{
+			name:   "missing operation",
+			filter: "(attr-value)",
+			err:    ErrMissingOperation,
+		},
+		{
+			name:   "invalid attribute name",
+			filter: "(attr$=value)",
+			err:    ErrInvalidAttrName,
+		},
+		{
+			name:   "missing end",
+			filter: "(attr=value",
+			err:    ErrUnexpectedEOF,
+		},
+		{
+			name:   "extra chars",
+			filter: "(attr=value)X",
+			err:    ErrUnexpectedInput,
+		},
+		{
+			name:   "unexpected non-filter",
+			filter: "(&(attr=value)X)",
+			err:    ErrUnexpectedInput,
+		},
+		{
+			name:   "no subfilters",
+			filter: "(&)",
+			err:    ErrUnexpectedInput,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := Parse(tt.filter)
+			require.ErrorIs(t, err, tt.err)
+		})
+	}
+}
+
+func Test_FilterParse_Unimplemented(t *testing.T) {
+	tests := []struct {
+		name   string
+		filter string
+		err    error
+	}{
+		{
+			name:   "less than or equal",
+			filter: "(attr<=value)",
+			err:    ErrUnimplemented,
+		},
+		{
+			name:   "greater than or equal",
+			filter: "(attr>=value)",
+			err:    ErrUnimplemented,
+		},
+		{
+			name:   "tilde equal",
+			filter: "(attr~=value)",
+			err:    ErrUnimplemented,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := Parse(tt.filter)
+			require.ErrorIs(t, err, tt.err)
+		})
+	}
+}
+
+func Test_FilterMatch(t *testing.T) {
+	attrs := map[string]any{
+		"dn":          "dc=example,dc=com",
+		"objectClass": "top",
+		"uid":         "1234",
+	}
+	e, err := NewEntryFromMap(attrs)
+	require.NoError(t, err)
+
+	tests := []struct {
```

Comment:
there is something about this table driven test structure that i don't like - but i don't have answers. it is usually:

```
Test_HEADLINE_NAME
// loads of similar- ish test cases that use up more than one screen hight
// the actual test t.Run and require.Equal.... (when reading this i've already forgotten the headline and need to scroll and then forget again).
```
I wish it could be something like:

```
Test_HEADLINE_NAME
// one simple test
// the actual test t.Run and require.Equal....

// all the other similar-ish test cases or so....
```

just a thing I think often....